scheduler/pkg/server/server_test.go

```diff
@@ -53,94 +53,265 @@ func TestLoadModel(t *testing.T) {
-				{ServerName: "server1", ReplicaIdx: 1, Shared: true, AvailableMemoryBytes: 1000,
-					ReplicaConfig: &pba.ReplicaConfig{InferenceSvc: "server1", InferenceHttpPort: 1, MemoryBytes: 1000, Capabilities: []string{"sklearn"}}}},
-			model: &pb.Model{Meta: &pb.MetaData{Name: "model1"}, ModelSpec: &pb.ModelSpec{Uri: "gs://model", Requirements: []string{"sklearn"}, MemoryBytes: &smallMemory}, DeploymentSpec: &pb.DeploymentSpec{Replicas: 2}},
-			code:  codes.OK},
+				{
+					ServerName:           "server1",
+					ReplicaIdx:           0,
+					Shared:               true,
+					AvailableMemoryBytes: 1000,
+					ReplicaConfig: &pba.ReplicaConfig{
+						InferenceSvc:      "server1",
+						InferenceHttpPort: 1,
+						MemoryBytes:       1000,
+						Capabilities:      []string{"sklearn"},
+					},
+				},
+				{
+					ServerName:           "server1",
+					ReplicaIdx:           1,
+					Shared:               true,
+					AvailableMemoryBytes: 1000,
+					ReplicaConfig: &pba.ReplicaConfig{
+						InferenceSvc:      "server1",
+						InferenceHttpPort: 1,
+						MemoryBytes:       1000,
+						Capabilities:      []string{"sklearn"},
+					},
+				},
+			},
+			model: &pb.Model{
+				Meta: &pb.MetaData{Name: "model1"},
+				ModelSpec: &pb.ModelSpec{
+					Uri:          "gs://model",
+					Requirements: []string{"sklearn"},
+					MemoryBytes:  &smallMemory,
+				},
+				DeploymentSpec: &pb.DeploymentSpec{Replicas: 2},
+			},
+			code: codes.OK,
+		},
 		{
 			name: "TwoReplicasFail",
 			req: []*pba.AgentSubscribeRequest{
-				{ServerName: "server1", ReplicaIdx: 0, Shared: true, AvailableMemoryBytes: 1000,
-					ReplicaConfig: &pba.ReplicaConfig{InferenceSvc: "server1", InferenceHttpPort: 1, MemoryBytes: 1000, Capabilities: []string{"sklearn"}}},
-				{ServerName: "server1", ReplicaIdx: 1, Shared: true, AvailableMemoryBytes: 1000,
-					ReplicaConfig: &pba.ReplicaConfig{InferenceSvc: "server1", InferenceHttpPort: 1, MemoryBytes: 1000, Capabilities: []string{"foo"}}}},
-			model: &pb.Model{Meta: &pb.MetaData{Name: "model1"}, ModelSpec: &pb.ModelSpec{Uri: "gs://model", Requirements: []string{"sklearn"}, MemoryBytes: &smallMemory}, DeploymentSpec: &pb.DeploymentSpec{Replicas: 2}},
-			code:  codes.FailedPrecondition}, // schedule to 2 replicas but 1 fails
+				{
+					ServerName:           "server1",
+					ReplicaIdx:           0,
+					Shared:               true,
+					AvailableMemoryBytes: 1000,
+					ReplicaConfig: &pba.ReplicaConfig{InferenceSvc: "server1",
+						InferenceHttpPort: 1,
+						MemoryBytes:       1000,
+						Capabilities:      []string{"sklearn"},
+					},
+				},
+				{
+					ServerName:           "server1",
+					ReplicaIdx:           1,
+					Shared:               true,
+					AvailableMemoryBytes: 1000,
+					ReplicaConfig: &pba.ReplicaConfig{InferenceSvc: "server1",
+						InferenceHttpPort: 1,
+						MemoryBytes:       1000,
+						Capabilities:      []string{"foo"},
```

Comment:
:information_source: :wrench: This test is currently failing because it breaks a key assumption of the source code: that server replicas have identical capabilities.  This is a reasonable assumption -- guaranteed in k8s in fact, because of how a `Server` resource is defined -- and sensible outside k8s.  If server replicas have different capabilities and memory/overcommit capacities, then they should really be different server resources.

I realise that this is hard to _enforce_ outside k8s for the time being, but I'd argue anyone not adhering to the above assumptions is implementing anti-patterns that make server administration difficult.

In a follow-up PR to set capabilities per server, there could be validation logic that ensures this assumption is met, and otherwise rejects the agent attempting to register with the scheduler.

Reply:
The same applies to `scheduler_test.go` and its failing `NotEnoughReplicas` test.  This test case is also forcing a failure by setting non-matching capabilities for a replica, even though this breaks an assumption.  Really these should either mock the underlying calls to inject failures, or should use cases where server replicas are otherwise incompatible, such as a lack of available memory.

Reply:
That does make sense. I think having validation (in a follow-up) is key. Are there documentation changes we could make too?

Reply:
Documentation changes are a good suggestion.  While I think it's best to enforce any assumptions, e.g. by rejecting the registration of a new agent if it makes claims that diverge from other replicas in the server it implements, clarifying expectations in docs is no bad thing.  For the sake of expediency and this PR not growing any more, I'll make any docs changes in a follow-up PR.